fix(instrumentation-fetch): release HTTP connection when response body is cancelled

Lei-k opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/main/guides/contributor#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

### What
When using `FetchInstrumentation`, if a user cancels the `Response.body` reader, the underlying HTTP connection is not released.

### Why
The instrumentation clones the `Response` and continuously reads from the clone to track span completion. Even if the user cancels the original response, the clone reader continues reading, preventing the browser from freeing the HTTP connection. This breaks integrations such as `flv.js` where `player.destroy()` is expected to close network connections.

Fixes #5883 

## Short description of the changes

Introduce a `withCancelPropagation` helper that wraps the original response body. When the user cancels, the cancel signal is propagated to the clone's reader, ensuring both readers stop and the HTTP connection is released properly.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

1. Start a `flv.js` player with `@opentelemetry/instrumentation-fetch` enabled.
2. Call `player.destroy()`.
3. Verify in Chrome DevTools that the HTTP request is marked as finished and the connection is released.

## Checklist:

- [x] Followed the style guidelines of this project

Lei-k commented:
All tests passed locally:

> Lerna (powered by Nx) Successfully ran target test for 37 projects (2m)

pichlermarc reviewed (commented):
Thank you for working on this! Could you please add a test for us to see that the behavior is correct now and also for us to prevent any regressions in the future :)

Lei-k commented:
Hi @pichlermarc,

I've added a test case that simulates `fetch` in a long connection scenario where `reader.cancel()` should properly release the connection. The test case is outlined below:

1. Use a continuously pushing `ReadableStream` to simulate a long connection
2. After calling `fetch`, read the first `chunk` and confirm the stream hasn't ended
3. After calling `reader.cancel()`, check if `traceFetch()` completes within one second. If it doesn't complete, it indicates that `cloneReader.cancel()` wasn't triggered to clean up resources

I'm testing this way because `cloneRes` exists internally within `FetchInstrumentation`, and there's no direct way to detect whether `cloneRes` has ended, so this approach is used to handle the situation.

legendecas commented on experimental/packages/opentelemetry-instrumentation-fetch/src/fetch.ts:
```diff
@@ -451,16 +450,69 @@ export class FetchInstrumentation extends InstrumentationBase<FetchInstrumentati
+          let proxiedResponse: Response | null = null;
+
           try {
             const resClone = response.clone();
```
This clone is redundant and can be removed if we are going to create a `Response` on line 510.

Lei-k replied:
Hi @legendecas  , I tried moving `endSpanOnSuccess`/`endSpanOnError` into the ReadableStream's controller inside `withCancelPropagation`. That works and would let us drop resClone.

However, it changes FetchInstrumentation’s semantics: today we eagerly start consuming the body as soon as the instrumentation is set up, regardless of whether the caller reads. Our existing `fetch.test.ts` relies on that eager behavior. When the callbacks live in the controller, the body is only consumed if/when the consumer reads, which causes many tests to fail (and the span never complete if the body isn’t read).

To preserve current behavior and avoid a large test break in this PR, I’d prefer to keep `resClone`. In a follow-up, we can switch to the consumer-driven model (end span on body read) and update the test suite to reflect the new semantics.

Lei-k commented:
Hi @legendecas , I’ve addressed the feedback. Could you please let me know if this PR meets the criteria for merging, or if there’s anything else you’d like me to change?
